### PR TITLE
Remove libsnark, libgmp, mini-gmp

### PR DESCRIPTION
**Pull request by DeckerSU**
#### Changes

- The old `mini-gmp` C-library has been removed, and the functions `bitcoin_base58{en,de}code` have been replaced with internal `{En,De}codeBase58` from the Bitcoin codebase.
- GMP functions are no longer used in `rewards CC`; the same results can be achieved more efficiently using the internal `arith_uint256` type.
- GMP functions are no longer used in `payments CC`; refactoring has been done to work with large numbers in calculations using `arith_uint256`.
- `Dilithium` and `musig` have been removed from the CC in preparation for the upcoming update to the `secp256k1` codebase.
- Old pre-sapling Sprout (PHGRProof) proofs validation has been removed. `komodod` will no longer verify old Sprout proofs and will assume they are valid. This change is safe for `KMD` and ACs, as transactions involving Sprout after `KOMODO_SAPLING_DEADLINE` (February 15, 2019) were prohibited at the consensus level.
- `libnsnark` and `libgmp` have been removed from dependencies and build systems.
- KMD and dPoW-protected AC checkpoints have been updated.

These changes will result in faster Initial Block Download (IBD) due to saving CPU time on verifying old Sprout transactions and disabling scripts and other resource-intensive checks during IBD. Additionally, with the removal of old pre-sapling Sprout, the need for `sprout-proving.key` and `sprout-verifying.key` files in `.zcash-params` (~910 Mb) is eliminated.

#### Sources

- Corresponding PR in KomodoOcean codebase: https://github.com/DeckerSU/KomodoOcean/pull/70
- https://github.com/zcash/zcash/pull/4060

#### Todo for review

I have already completed all the tasks on the TODO list by myself, but it would be beneficial to have them reviewed by someone else (please regard this list as key points or guidelines for what needs to be checked):

- [ ] Compare the changes in the `KomodoOcean` and original `komodod` sources related to this pull request file by file to ensure that we have all the necessary changes in both codebases and have not missed or forgotten anything.
- [ ] Test the synchronization of all dPoW-protected chains (ACs) and `KMD` itself from the beginning. If feasible, compare the synchronization time before and after the changes made in this pull request.
- [ ] Review the changes with a focus on potential attacks, such as attempting to manipulate nodes during initial block synchronization with fake chains containing old proofs, creating invalid transactions, or trying to violate the established rules.
- [ ] Conduct an additional check on the logic for calculating large numbers in the Rewards CC and Payments CC.
- [ ] Test the build, run, and functionality of the daemon on different platforms (Linux, Windows, MacOS), even though we already have a build and other tests on the CI/CD system.
- [ ] Consider adding to the documentation a note that new AC chains must include the `-ac_sapling=1` parameter to support private transactions from the beginning. Otherwise, these chains will not be able to send shielded transactions until sapling activation, as old Sprout proofs are no longer supported and additionally, all Sprout transactions have been prohibited since February 15, 2019 by consensus rules.
- [ ] Verify that all functions operate correctly without the presence of the `sprout-proving.key` and `sprout-verifying.key` files in the Zcash parameters.

### CODE REVIEW

**Comment by smk762** [comment]
Should this also be removed? https://github.com/KomodoPlatform/komodo/blob/patch-remove-libsnark/libsnark.mk.patch

I notice there are no updates to readme, so I assume listed deps are valid, if not I'll PR changes as required.

**Comment by TheComputerGenie** [comment]
> Should this also be removed? https://github.com/KomodoPlatform/komodo/blob/patch-remove-libsnark/libsnark.mk.patch
> 
> I notice there are no updates to readme, so I assume listed deps are valid, if not I'll PR changes as required.

All of those patches from 2017 should be dumped

File: zcutil/fetch-params.bat [hunk 1]
```diff
@@ -4,14 +4,14 @@ cd %THIS_DIR%
...
+@REM IF NOT EXIST "%APPDATA%"\ZcashParams\sprout-proving.key (
+@REM     ECHO Downloading Zcash trusted setup sprout-proving.key, this may take a while ...
+@REM     .\wget64.exe --progress=dot:giga --continue --retry-connrefused --waitretry=3 --timeout=30 https://z.cash/downloads/sprout-proving.key.deprecated-sworn-elves -O "%APPDATA%"\ZcashParams\sprout-proving.key
+@REM )
+@REM IF NOT EXIST "%APPDATA%"\ZcashParams\sprout-verifying.key (
+@REM     ECHO Downloading Zcash trusted setup sprout-verifying.key, this may take a while ...
+@REM     .\wget64.exe --progress=dot:giga --continue --retry-connrefused --waitretry=3 --timeout=30 https://z.cash/downloads/sprout-verifying.key -O "%APPDATA%"\ZcashParams\sprout-verifying.key
+@REM )
 IF NOT EXIST "%APPDATA%"\ZcashParams\sapling-spend.params (
     ECHO Downloading Zcash trusted setup sprout-proving.key, this may take a while ...
```
Comment (dimxy) [inline zcutil/fetch-params.bat:16]: I guess not sprout-proving but sapling-spend should be here in echo msg?

Comment (DeckerSU) [reply zcutil/fetch-params.bat]: > I guess not sprout-proving but sapling-spend should be here in echo msg?

Yes, true. Echo messages fixed.

**Comment by DeckerSU** [comment]
> Should this also be removed? https://github.com/KomodoPlatform/komodo/blob/patch-remove-libsnark/libsnark.mk.patch

Yes, not only this file, but also all `*.patch` files in the repository root were leftover from previous operations unrelated to this pull request. They were accidentally left in the repository but have now been removed.

**Comment by DeckerSU** [comment]
@dimxy Also, I propose removing all dynamic (time-based) sapling activation mechanisms from our codebase due to the complications they introduce, such as the `futureblock` flag and its unclear logic. Instead, I suggest hardcoding the known sapling activation heights for dPoW protected chains and setting sapling activation to begin at `height.1` for new AC chains. This approach will eliminate the need for the "tricks" implemented during the sapling transition period, ensuring a cleaner codebase and facilitating the integration of future updates. If you agree with this proposal, I will create a separate issue and related pull request for further discussion.

**Comment by dimxy** [comment]
> @dimxy Also, I propose removing all dynamic (time-based) sapling activation mechanisms from our codebase due to the complications they introduce, such as the `futureblock` flag and its unclear logic. 

I am okay with this idea. I think it is also good to create vUpgrades arrays for asset chains, like make a dedicated vUpgrades array for each old known chain and a default array for new chains with special upgrades.
(Btw as far as I remember futureblock has several purposes, not only related to sapling code)

**Pull request merged**